insonmnia/npp/relay/server.go

```diff
@@ -134,6 +237,52 @@ func (m *connPool) popRandom() *meeting {
 	return v
 }
 
+type meetingHandler struct {
+	bufferSize int
+	metrics    *netMetrics
+	log        *zap.SugaredLogger
+}
+
+func (m *meetingHandler) Relay(ctx context.Context, server, client net.Conn) error {
+	log := m.log.With(zap.Stringer("server", server.RemoteAddr()), zap.Stringer("client", client.RemoteAddr()))
+	log.Info("ready for relaying")
+	defer log.Info("finished relaying")
+
+	wg := errgroup.Group{}
+	wg.Go(func() error {
+		return m.transmitTCP(server, client, m.metrics.TxBytes, log)
+	})
+	wg.Go(func() error {
+		return m.transmitTCP(client, server, m.metrics.RxBytes, log)
+	})
+
+	return wg.Wait()
+}
+
+func (m *meetingHandler) transmitTCP(from, to net.Conn, metrics *atomic.Uint64, log *zap.SugaredLogger) error {
+	buf := make([]byte, m.bufferSize)
+
+	for {
+		bytesRead, err := from.Read(buf[:])
```

Comment:
AFAIK, bytesRead can be non-zero even in case non-nil err

Reply:
Fixed in #582, because it's out of the scope (just code moving here).